plugins/module_utils/foreman_helper.py

```diff
@@ -1094,6 +1097,15 @@ def _update_entity(self, resource, desired_entity, current_entity, params, forem
                     sort_key = 'name'
                 else:
                     sort_key = list(value[0].keys())[0]
+                # special handling for parameters created by ParametersMixin
```

Comment:
@mdellweg how much of a decontamination shower do you need after reading this?

Reply:
also should probably move that up by 4 lines so it's not in the middle of the sorting logic

Reply:
![image](https://github.com/user-attachments/assets/8141ed16-96b0-4778-9872-20527307d989)

Oh dear. Part of me is wondering if this should be handled by the mixin somehow. But i guess that would be a huge architectural change now.
Also I'm a bit confused:
1. The use of `value` and `new_value` seems mixed up here.
2. Should we not drop all the list entries that are the same in `new_value` and `old_value` at this point?

Reply:
> Part of me is wondering if this should be handled by the mixin somehow. But i guess that would be a huge architectural change now.

I tried, but couldn't come up with a solution that would actually work.
Can certainly try again, now that most nightmares of the first implementation round are over.

>     1. The use of `value` and `new_value` seems mixed up here.

How so?

The idea was:
1. take the `new_value` (aka: the parameters the user has set in their playbook) and iterate over it (thus, looking at each parameter individually)
2. find the same parameter in the `old_value` (aka: the one known to the server)
3. when found, remove all parameter-attributes that are set in the old but not the new version of the parameter
4. let the normal flow of the `update_entity` compare the values.

>     2. Should we not drop all the list entries that are the same in `new_value` and `old_value` at this point?

You mean `old_param` and `new_param`?
We could, but I don't think there is a big benefit. Actually, I think this might break things when we don't pass at least `name` and `value` -- no idea how the API behaves then.

Reply:
> We create a (shallow) copy `new_value` of `value` in line 1086 in case we want to overwrite it. But we still use `value` in line 1095 for example.
> The shallow here might be an actual issue when we pop anything from within `new_value`, for `value` being the thing ansible will report in the details about the modules use later.

Yeah. I think the idea was (but the shallow copy slightly breaks that): we use `new_value` for all the comparison magic (including partially converting things, because, Ansible), but *iff* we decided that the new value is actually new and needs to be submitted, we take the *raw* `value` that the user provided.

> What if the user specified one changed and one unchanged parameter? Will we send both to the update call? Should we?

We have to. Otherwise the API would delete the unsent parameter. Or so I expect it to.



> For better understanding, the old_value looks like this?

Yes